IIP-0011: Core Move View Functions (#43)
iotaledger/IIPs: iips/IIP-0011/IIP-0011.md

---
iip: 11
title: Core Move View Functions
description: Move view functions - definition and specification at language level
author: Lorenzo Benetollo (@lollobene) <lorenzo.benetollo@iota.org>, Mirko Zichichi (@miker83z) <mirko.zichichi@iota.org>
discussions-to: https://github.com/iotaledger/IIPs/discussions/42
status: Draft
type: Standards Track
layer (*only required for Standards Track): Core
created: 2026-04-27

> requires (*optional): None

Reviewer: requires package metadata

replaces (*optional): None
---

## Abstract

This proposal introduces _view functions_ in Move, a class of functions that are guaranteed to not modify the persistent global storage. A view function is explicitly annotated using `#[view]` and its properties are statically enforced by the compiler.

These functions are suitable for reading from the ledger state and/or executing code in Move with no effect on the state. This proposal, then, aims to improve developer ergonomics, enable safer APIs, and allow optimized execution paths for read-only on-chain queries.

## Motivation

Smart contract developers frequently need to query on-chain storage without modifying it. Today, this is harder than it should be:

> - No distinction between read-only and state-modifying functions. If your dApp depends on a third-party Move package, you have no way to know whether calling a function will mutate state without manually auditing its code. This increases the risk of [unintended storage mutations](https://olivine-hydrofoil-637.notion.site/Nemo-Security-Incident-Cause-Process-and-Fund-Tracing-Report-V1-1-26a6b8723d8a80e29cb8cb48fe1390f2 "https://olivine-hydrofoil-637.notion.site/Nemo-Security-Incident-Cause-Process-and-Fund-Tracing-Report-V1-1-26a6b8723d8a80e29cb8cb48fe1390f2")

Reviewer: I'd place the link behind an explicit "see Nemo Protocol security incident"

- No clear or lightweight way to read on-chain data. If you deploy a Move package with custom object types, reading their fields requires building a custom indexer, a heavyweight solution for what should be a simple query.

>   Introducing view functions provides:

Reviewer: Need a blank space above to fix the indentation

- A clear and **statically enforced** guarantee of read-only behavior
- Improved developer experience and reduced risk of unintended state mutations
- Opportunities for optimized execution and seamless [integration with RPC](https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0005/iip-0005.md "https://github.com/iotaledger/IIPs/blob/main/iips/IIP-0005/iip-0005.md") interfaces and query-oriented tooling, enabling off-chain evaluation without gas costs or signature requirements

## Background

> In Move on IOTA*,* the original [Move Resource model](https://developers.diem.com/papers/diem-move-a-language-with-programmable-resources/2019-06-18.pdf "https://developers.diem.com/papers/diem-move-a-language-with-programmable-resources/2019-06-18.pdf") has evolved into an **object-centric** model, in which structs with the `key` ability and a required `UID` (as a first field) represent on-chain objects.

Reviewer: ```suggestion
In Move on IOTA, the original [Move Resource model](https://developers.diem.com/papers/diem-move-a-language-with-programmable-resources/2019-06-18.pdf "https://developers.diem.com/papers/diem-move-a-language-with-programmable-resources/2019-06-18.pdf") has evolved into an **object-centric** model, in which structs with the `key` ability and a required `UID` (as a first field) represent on-chain objects.
```

As a consequence, functions must **explicitly declare as parameters all the objects they operate on**, either by value or by reference, while the virtual machine is responsible for loading and deserializing these objects at execution time.

This eliminates implicit global storage access and requires all storage interactions to occur via **explicitly** **passed** **objects**.

> ```move

Reviewer: I'd remove this source code since is shown just below

public struct Object has key { id: UID, val: u64 }

public fun create(addr: address, ctx: &mut TxContext) {
    let o = Object { id: object::new(ctx), val: 44 };
    transfer::transfer(o, addr);
}

public fun edit(o: &mut Object, new_val: u64) { o.val = new_val; }

public fun delete(o: Object) { let Object { id, val } = o; object::delete(id); }
```

> In _IOTA Move_, object creation is not exposed as a language-level primitive.

Reviewer: ```suggestion
Additionally, in Move on IOTA, object creation is not exposed as a language-level primitive.
```

Instead, it is realized through framework-provided functions (e.g., `object::new`) combined with transfer operations.

**This design implies that object creation is not performed by any language primitive but is delegated to the platform framework. More generally, as we will see in the next section, the storage can be modified either by language primitives or by framework functions.**

> ## What can alter the storage?

Reviewer: I'd suggest to use this hierarchy:
`## Background`
`###  What can alter the storage?`

As previously mentioned, in Move on IOTA, objects are the only entities stored in persistent storage. Consequently, all changes to the global storage occur through operations on objects.

These operations include:

- **Creating an object**, which adds a new object to the storage

  ```move
  public struct Object has key, store { id: UID, value: u64 }
  public struct Wrapper has key { id: UID, o: Object }

  public fun create(addr: address, ctx: &mut TxContext) {
      let o = Object { id: object::new(ctx), val: 44 };
      transfer::transfer(o, addr);
  }
  ```

- **Modifying an object’s fields**, which updates its internal state

  ```move

>   public fun edit(o:  &mut  Object, new_val:  u64)  { o.val = new_val; }

Reviewer: indentation for all functions must be fixed

  ```

- **Sharing an object**, which changes its accessibility model

  ```move
  public fun share(value:  u64, ctx:  &mut  TxContext)  { transfer::share_object(Object  { id:  object::new(ctx), value }) }
  ```

- **Wrapping an object**

  ```move
  public fun wrap(o:  Object, ctx:  &mut  TxContext)  { transfer::transfer(Wrapper  { id:  object::new(ctx), o }, ctx.sender()); }
  ```

- **Adding or removing dynamic fields**, which modify associated dynamic fields

  ```move
  public fun add_dynamic_field(self:  &mut  Object, value:  bool)  { dynamic_field::add(&mut  self.id,  b"dynamic_field", value); }
  ```

## Specification

### Definition

Based on the premise that global storage is solely constituted by objects in IOTA, then:

A _view function_ is a Move function that does not mutate any IOTA object and that returns a value, which is not an IOTA object.

### Rules

A compliant _view function_ is declared using the `#[view]` attribute, which enforces a static validation by the compiler:

- It MUST NOT create, update, or delete IOTA objects
- It MUST return a value. This value can be anything but an IOTA object _by value_\*.

> - It MAY read from the global storage through a reference to an object

Reviewer: ```suggestion
- It MAY read from the global storage through a reference to an object
- It MUST NOT have mutable references as parameters
```

- It MAY perform pure computation

* In this context, _by value_ refers to passing or returning a value in a way that transfers ownership of that value, rather than accessing it through a reference.

In Move, values passed by value are consumed by the callee, meaning that ownership is moved and the original binding can no longer be used. This is in contrast to references (`&T` or `&mut T`), which provide access to a value without transferring ownership.

> For view functions, returning or accepting objects by value is disallowed, as it would imply ownership transfer or potential modification of persistent state.

Reviewer: All of this should go into the background

### Annotation

A view function is declared as:

```move

> #[view] public fun foo(...): T { ... }

Reviewer: indentation

Therefore, even if a view function calls other functions, those calls are effectively restricted to functions whose parameter and return types are compatible with view constraints. This ensures that state-modifying operations cannot be reached, even transitively.

### Invocation of View Functions

Any other function MAY invoke a function annotated with `#[view]`, so the constraints imposed on view functions MUST hold independently of the caller context.

Although a view function may not receive objects by value or as mutable references, allowing mutable references to **non-object** types would still enable indirect state modifications. For example, a mutable reference to an object’s field could be passed to a view function, allowing modification of the object’s field through that reference.

To **prevent** such **indirect mutations**, view functions MUST NOT accept mutable references (`&mut T`) as parameters, regardless of whether `T` is an object type.

For example:

```move
public fun edit(mut o: Object, new_val: u64): Object {
        edit_u64_by_ref(&mut o.val, new_val);
        o

>     } // NON view function because it can mutate an object field

Reviewer: indentation

Author: what happened here

    public fun edit_u64_by_ref(val: &mut u64, new_val: u64): u64 {
        let old_val = *val;
        *val = new_val;
        old_val
    }
```

### **Function Signature Constraints**

A function annotated with `#[view]` MUST satisfy:

- **Visibility**

>   - MUST be `public` or `public entry`

Reviewer 2: Why does it have to be `public`? What is wrong with private view functions?

Author: I guess it comes from this idea 

> a standardized interface for enabling ergonomic and application-specific queries to on-chain state
 
from IIP-0005. So `private` would not make sense?

- **Return values**
  - MUST return values
  - MUST NOT return IOTA objects
- **Parameters**
  - MUST NOT include IOTA objects passed by value
  - MAY include:
    - immutable references (`&T`)

>     - primitive types or user-defined types with `Copy` or `Drop` ability (**non IOTA objects, nor types that can contain objects**)

Reviewer: ```suggestion
    - primitive types or user-defined types with `Copy` or `Drop` ability (**non IOTA objects, nor types that can contain objects**) passed by value
```

Reviewer: Maybe we could change all the occurrences of by value to by-value

Author: not everywhere, only if it is used as an adjective

  - **References**
    - MUST NOT use mutable references (`&mut T`) to any type, including `TxContext`
    - MAY use immutable references
- **Type Parameters**

>   - MUST be types with `Copy` or `Drop` ability (this ensures are not objects)

Reviewer: ```suggestion
  - MUST be types with at least one among `Copy` and `Drop` ability (this ensures are not objects)
```

        s
    }

    // Error: mutable reference
    #[view]
    public fun edit_u64_by_ref(val: &mut u64, new_val: u64): u64 {
        let old_val = *val;
        *val = new_val;
        old_val
    }

    // Error: Wrapper type contains an object type
    #[view]
    public fun delete_wrapped_object(wrapper: Wrapper): u64 {
        let Wrapper { o } = wrapper;
        let S { id, val } = o;
        object::delete(id);
        val
    }
}
```

## **Rationale**

This proposal introduces an explicit abstraction for read-only execution, similar to “view” or “pure” functions in other smart contract platforms.

Key design decisions:

- **Explicit annotation (**`#[view]`**)**
  - Avoids ambiguity and enables static guarantees
- **Static enforcement**
  - Ensures safety at compile time
- **Recursive restriction**
  - Guarantees no indirect storage mutation
- **Separation of constraints**
  - Improves clarity and maintainability

## **Backwards Compatibility**

This proposal is backwards compatible:

- Existing code is unaffected
- The `#[view]` attribute is optional

However:

- Adding `#[view]` introduces stricter constraints that are enforced at compile time

> - Removing `#[view]` may affect client expectations, particularly for off-chain execution and API usage

Reviewer 2: Can you or should you be able to remove the #[view] annotation in an upgrade?

Author: Good point, I would not allow it. What do you think?

Reviewer: With an upgrade a public function cannot change signature, a private one can. It depends if a view function is allowed to be private (see above)

Author: so do we agree on view functions not being private?

> As an implementation detail, view functions MAY be included in `PackageMetadata`, similarly to how authenticator functions are currently exposed. This would allow clients and tooling to reliably discover view functions without requiring additional static analysis.

Reviewer: IS included

Reviewer: In this section, we should put a reference to the iip 10 and iip 5 (proposal in 5 will be made compatible with this proposal).

  }
  ```

- **Modifying an object’s fields**, which updates its internal state

  ```move
  public fun edit(o:  &mut  Object, new_val:  u64) {
      o.val = new_val;
  }
  ```

- **Deleting an object**, which removes it from storage

  ```move
  public fun delete(o:  Object) {
      let  Object  { id, val }  = o;
      object::delete(id);
  }
  ```

- **Transferring an object**, which changes its ownership

  ```move
  public fun transfer(o:  Object, recipient: address) {
      transfer::public_transfer(o, recipient)
  }
  ```

- **Sharing an object**, which changes its accessibility model

  ```move
  public fun share(value:  u64, ctx:  &mut  TxContext) {
      transfer::share_object(Object  { id:  object::new(ctx), value })
  }
  ```

- **Wrapping an object**

  ```move
  public fun wrap(o:  Object, ctx:  &mut  TxContext) {
      transfer::transfer(Wrapper  { id:  object::new(ctx), o }, ctx.sender());
  }
  ```

- **Adding or removing dynamic fields**, which modify associated dynamic fields

  ```move
  public fun add_dynamic_field(self:  &mut  Object, value:  bool) {
      dynamic_field::add(&mut  self.id,  b"dynamic_field", value);
  }
  ```

### By-value semantics

> In this proposal, a value is passed or returned _by value_ when ownership of that value is transferred, rather than when the value is accessed through a reference.

Reviewer 2: Can we name it differently? For example: a resource/variable is passed or returned by value...
It is hard to read when both are called value.

Author: ```suggestion
In this proposal, an argument or return value, whether an object or a native type, is considered to be passed _by value_ when ownership is transferred, rather than when access is provided through a reference.
```

> In Move, passing a value by value consumes it: ownership moves to the callee, and the original binding can no longer be used. References (`&T` and `&mut T`) instead provide access to a value without transferring ownership.

Author: ```suggestion
In Move, passing an argument by value consumes it: ownership moves to the callee, and the original binding can no longer be used. References (`&T` and `&mut T`) instead provide access to a value without transferring ownership.
```

This distinction matters for view functions because accepting or returning an IOTA object by value would move ownership of a persistent object, which can enable ownership changes, wrapping, deletion, or other state-modifying behavior. For this reason, view functions cannot accept or return IOTA objects by value.

## Specification

### Definition

Based on the premise that global storage is solely constituted by objects in IOTA, then:

> A _view function_ is a Move function that does not mutate any IOTA object and that returns a value, which is not an IOTA object or a reference.

Reviewer 2: Reading the IIP down below makes me think we can't return read-only references to objects either, is that right?
Shouldn't we allow that? Can we return read-only references of structs/pure values? If not, what is the problem with it?

Author: Yes, we currently disallow returning immutable references because they cannot cross the execution boundary as references. We could model this as implicitly reading the reference and returning the serialized value, but that only works for copyable types and would make `&T` behave like `T` at the API boundary, which is misleading and not sound

Reviewer 2: > they cannot cross the execution boundary as references.

by boundary you mean
 a, end of the current frame (view function),
 b, end of the current PTB command,
 c, or end of execution, so the RPC boundary?

I suspect you meant `c`, but maybe I misunderstood. In this case we are at the RPC layer so we don't need to care about type abilities. You can read any object in the public ledger off-chain anyway and "copy" their content, it's just not in the move world.

In case you meant `b`, we will have reference passing in PTBs in the future, so making this restriction just because we don't have that feature implemented yet is not a showstopper. If you compose view functions in move code it would already work, no?

Author: Yes, I meant both `b` and `c`.
In case of `b`, if PTB reference passing will be allowed, then this is no longer a fundamental limitation, so a returned `&T` could be passed across PTB commands.
For `c`, I agree the RPC can return a serialized snapshot of the referenced value. So, for example, the RPC would read `&T` and expose `T`.
